### ISSUE TITLE
ODE models behave differently from normal models: 'A' parameter lacking default value

### ISSUE THREAD

**Issue by Pitje06** [tBuLi/symfit#63]
ODE models behave differently that normal models, I expect the following snippet to work, and give me values for `A`:
```python3
A, t = variables('A, t')
k = Parameter()
model = ODEModel({D(A, t): - k * A}, initial={t: tdata[0], A: concentration[0]})

print(model[A](k=1, t=tdata))  # 'A' parameter lacking default value
```
I use `A` as index, because it is an dependant variable according to the model object.
But instead I get the following error: TypeError: 'A' parameter lacking default value.

Calling `print(model(k=1, t=tdata))` does work as expected.

This is reason why PR #33 currently does not work with ODE models.

**Comment by tBuLi** [comment]
What happens if you do 
```python
model(k=1, t=tdata)
```
instead of 
```python
model[A](k=1, t=tdata)
```

That is the recommended way of calling models. Unpack after the fact, because the two are not equivalent.

You're right that it would be nice if your way would work as well, but I think it will be fundamentally impossible. I will look into this asap.

**Comment by Pitje06** [comment]
I could do that, but then I'd need to find the proper index associated with A in the answer I get out (No, I can't use A as index for some reason). But then I'd need to rework a rather annoying amount of code. I will if needed, but I'd really rather not.

It's unfortunate that a model is a model is not a model.

**Comment by tBuLi** [comment]
Is ``range(0, 10)[0]`` a ``range``? The behavior upon calling a ``Model`` is guaranteed, but ```model[A](k=1, t=tdata)``` does not call a model. It returns a symbolic component, which is then called with the provided arguments.

For a normal model, we have ``y_i = f_i(x, ...)`` for every component. For a system of ODEs we have ``d(y_i)/dx = f_i(y_1, ..., y_m, x, ...)``. Indexing any ``Model`` returns the symbolic expression for that component and therefore for an ``ODEModel`` it should return ``f_i(y_1, ..., y_m, x, ...)`` (a symbolic expression).

That is why in general ```model[A](k=1, t=tdata)``` simply can't work. For given values of ``y_1, ..., y_m, x, ...`` the symbolic expression ``model[A]`` can be evaluated. But that is off course not what we want here, you want the numerical value of ``y_i``. Therefore the model has to be called through the normal API so that the system of ODEs can be integrated.

Bottom line: indexing models returns the symbolic expression for that component, not a numerical expression. This is fundamental to the whole structure of ``symfit``. They should not be called directly if you need numerical results.

So I really think this should be rewritten to the recommended syntax. I get that this might cause some difficulties, but the order of the return of the call is a ```namedtuple```. So I'm pretty sure this can be abused quite easily by doing 

```python
ans = model(k=1, t=tdata)
print(getattr(ans, A.name))
```

Furthermore, the order of this ``namedtuple`` is alphabetic so that might help.

**Comment by Pitje06** [comment]
Additional thought from my side made me agree with you before you posted ;) (it's a *system* of ODEs, so I can't evaluate just a part of it)
I'm reworking #33 to use the ~~preferred~~ correct syntax, including all my examples. It's not going to win any beauty-prizes, but it'll work. I hope.
